### ISSUE TITLE
Function name casing

### ISSUE THREAD

**Issue by erijo** [BlankSpruce/gersemi#21]
It would be nice if gersemi could be instructed to not always lower case function names, for cases where PascalCase is used instead of snake_case. E.g. official functions such as FetchContent_MakeAvailable() but also custom functions such as MyCoolFunction(). Either to maintain the casing used in the CMakeLists.txt or be able to enforce the casing used when the function is defined (or with a hint).

**Comment by BlankSpruce** [comment]
I will consider that but there are certain caveats to work out:
- Can I make it backwards compatible? I never liked the idea that upgrading formatter would require formatting changes except for inconsistency fixes.
- If I can make it backwards compatible
  - Can I do it in some manner that doesn't require extra work from user?
  - What's the best way to do that? 
    - `--maintain-cases`/configuration flag?
    - function signature hints?
    - only supporting official functions with this exception?
- Else
  - How can I do it in the least disruptive way?
  - What should I do with the code that uses official functions without "canonical" casing? From maintenance point of view I'd prefer idempotency but it might be not the least disruptive.
  - With change like below that leads to codebase reformatting, would it be obvious? I hope so, renaming is not a new thing but due to how CMake language states two names are equivalent the reformatting might be seen as unnecessary.
```diff
--- a/utils.cmake
+++ b/utils.cmake
@@ -1,4 +1,4 @@
-function(add_target)
+function(Add_Target)
     set(OPTIONS)
     set(ONE_VALUE_KEYWORDS NAME)
     set(MULTI_VALUE_KEYWORDS SOURCES DEPENDENCIES)
```

- Should I make this change backwards incompatible?
  - How can I help users understand what's going on?

If you have any reflections on these questions I'm all ears.

I don't know when I will implement it since I like to ponder on ideas like that to weigh all pros and cons. Especially recent issue about tabs vs spaces gave me new insight into the problem I didn't know beforehand. You can be certain though that this is a nice feature request that I think should be supported one way or another.

**Comment by erijo** [comment]
One possibility is to add something like `--function-case <lower|upper|no-change|canonical>` where lower would be the default (at least for now) and thus it would be backwards compatible. I agree that upgrading formatter shouldn't require (too many) formatting changes.

Without knowing how the code is structured, I assume that the first three are a bit easier to implement than the last. But getting support for "canonical" casing would be sweet as one of the goals with a formatter (IMO) is to get the same formatting independent on who it was that wrote the code.

It wouldn't even have to be the "canonical" casing. Being able to supply a list of function names and thus how they should be formatted would come a long way. Especially if it would be possible to decide what to do with names that aren't known (error, lower, no-change, etc.).

This "database" could perhaps be combined with `--function-case canonical` so that it can be used to override the default (canonical) casing or provide such a casing if the function isn't known (not a default cmake function nor a known custom one).

As for the renaming function case, I would see it as a feature if gersemi would update all callers for a function if the canonical name is updated (e.g. by renaming it as in your example).

**Comment by BlankSpruce** [comment]
I'm slowly moving my position to the following rule: "Use canonical name if it's official command or known custom command, otherwise lower case name". I'm opting for lower case instead of "no change" not because it's cheaper to implement for me but rather official CMake commands rather tend to have lower case names with a few exceptions like FetchContent_* family etc.

As for "database": I don't want to have another way of specifying known definitions because that's precisely against my philosophy about configuration knobs: the fewer and simpler, the better. If there are some commands which don't have their definition in your source tree, let's say external dependency, you can still provide stub definitions through `definitions` (command line or `.gersemirc`) in similar fashion as [mypy does it](https://mypy.readthedocs.io/en/latest/stubs.html) for Python. Example:

```yaml
# ./.gersemirc
definitions: [./src/cmake/stubs, ...] # ... other paths that might contain actual definitions
line_length: 120
list_expansion: favour-expansion
```

```cmake
# ./src/cmake/stubs/try_to_win_best_picture_academy_award.cmake
# A stub for some external command out of our control
function(try_to_win_best_picture_academy_award)
    # gersemi: hint { CAST: pairs, SUMMARY: command_line }
    set(options FOREIGN_LANGUAGE)
    set(oneValueArgs GENRE YEAR)
    set(multiValueArgs DIRECTORS CAST SUMMARY)

    cmake_parse_arguments(_ "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
endfunction()
```

It might feel a bit hacky but without a way to infer the intent from the source code that's the best I can manage. This approach avoids situation that there is more than one way to provide definition and I suspect usually most of custom commands will be under your project's control so stubs won't be needed for the majority of commands. Yes, this is error-prone to changes in external command definitions but so would be "database". CMake language heavily restricts sane solutions.

**Comment by erijo** [comment]
> I'm slowly moving my position to the following rule: "Use canonical name if it's official command or known custom command, otherwise lower case name".

Sounds great to me!

> As for "database": I don't want to have another way of specifying known definitions [...] you can still provide stub definitions

:+1: 

> I suspect usually most of custom commands will be under your project's control so stubs won't be needed for the majority of commands

Yes, that's true.

**Comment by BlankSpruce** [comment]
I've implemented most of the stuff to support this feature but the remaining part is going through list of official commands and adding their canonical name. Here's an odd thing:
`check_fortran_compiler_flag` in [official 3.29 doc](https://cmake.org/cmake/help/v3.29/module/CheckFortranCompilerFlag.html)
`CHECK_FORTRAN_COMPILER_FLAG` in [definition](https://gitlab.kitware.com/cmake/cmake/-/blob/50252733c46025d7086d7b0ea6daef04e0c5346f/Modules/CheckFortranCompilerFlag.cmake#L52)
`CHECK_Fortran_COMPILER_FLAG` in [test example](https://gitlab.kitware.com/cmake/cmake/-/blob/50252733c46025d7086d7b0ea6daef04e0c5346f/Tests/FortranOnly/CMakeLists.txt#L91)

You might say "we should follow the documentation". Check this out:
`check_fortran_compiler_flag` in [doc](https://cmake.org/cmake/help/v3.29/module/CheckFortranCompilerFlag.html)
`CHECK_FORTRAN_FUNCTION_EXISTS` in [doc](https://cmake.org/cmake/help/v3.29/module/CheckFortranFunctionExists.html)
`check_fortran_source_compiles` in [doc](https://cmake.org/cmake/help/v3.29/module/CheckFortranSourceCompiles.html)
`check_fortran_source_runs` in [doc](https://cmake.org/cmake/help/v3.29/module/CheckFortranSourceRuns.html)

`check_fortran_*` family doesn't have consistent casing. :neutral_face: 

I might have to put my thumb on the scales for _consistent consistent ™_ formatting.

**Comment by BlankSpruce** [comment]
@erijo 
If you were so kind to test the feature it's available [on this branch](https://github.com/BlankSpruce/gersemi/tree/canonical-casing). I've also extended [README about stub definitions](https://github.com/BlankSpruce/gersemi/tree/canonical-casing?tab=readme-ov-file#how-to-format-custom-commands-for-which-path-to-definition-cant-be-guaranteed-to-be-stable-eg-external-dependencies-not-managed-by-cmake) and in the [CHANGELOG](https://github.com/BlankSpruce/gersemi/blob/canonical-casing/CHANGELOG.md#changed) I've listed the commands for which I've made an exception to the rule about canonical casing like `check_fortran_function_exists`.

I'll release 0.13.0 either after your response or after a couple of days.

**Comment by erijo** [comment]
I've tested it now and AFAICT it works as expected. Thanks a lot!

**Comment by BlankSpruce** [comment]
I've released 0.13.0 along with fixes for other reported issues. Thanks for contribution.
